cupy/core/fusion.py

```diff
@@ -239,99 +344,192 @@ def copy(self):
-    if is_scalar or is_ndarray:
-        return mem.get_fresh(numpy.dtype(arg_type), const=arg)
-    raise Exception('Unsupported type %s' % arg_type)
-
-
-def _convert(f):
-    if type(f) is core.ufunc:
-        return _convert_from_ufunc(f)
-    if type(f) is core.ElementwiseKernel:
-        return _convert_from_elementwise(f)
-    raise Exception("Can't convert from %s to FusionOp" % type(f))
-
-
-def _should_use_min_scalar(in_args):
-    max_array_kind = -2
-    max_scalar_kind = -1
-    for i in in_args:
-        kind = _kind_score[i.ty.kind]
-        if i.const is None:
-            max_array_kind = max(max_array_kind, kind)
-        else:
-            max_scalar_kind = max(max_scalar_kind, kind)
-    return (max_scalar_kind != -1 and
-            max_array_kind >= max_scalar_kind)
+    Only when fusing the reduction, the following attributes are updated.
 
+        reduce_op (tuple): One of the element of reduction.***._raws._ops.
+        reduce_identity (any type): The identity value of the reduction.
+        reduce_kwargs (dict or None): kwargs of the reduction.
 
-def _convert_from_ufunc(ufunc):
-    nin = ufunc.nin
-    nout = ufunc.nout
+        premap_ret (_FusionVarCUDA or None): The target of reduction
+        postmap_param (_FusionVarCUDA or None): The result of reduction
+        postmap_op_list (list of FuisonOp): The post-map operations.
+        postmap_local_list (list of _FusionVarCUDA): The local variables which
+    appears in the post-map operations
+    """
 
-    def get_mem(args):
-        for i in args:
-            if type(i) == _FusionRef:
-                return i._mem
-        raise Exception('number of ndarray arguments must be more than 0')
+    def __init__(self):
+        self.preamble_set = set()
+        self.submodules = dict()
+        self.count = 0
 
-    def can_cast1(args, ty_ins):
-        for i in six.moves.range(nin):
-            if args[i].const is None:
-                if not numpy.can_cast(args[i].ty, ty_ins[i]):
-                    return False
+        self.op_list = []
+        self.param_list = []
+        self.local_list = []
+
+        self.reduce_op = None
+        self.reduce_identity = None
+        self.reduce_kwargs = None
+
+        self.postmap_op_list = []
+        self.premap_ret = None
+        self.postmap_param = None
+        self.postmap_local_list = []
+
+    def __repr__(self):
+        return '<_FusionMem, op_list={}, var_list={}>'.format(
+            self.op_list, self.var_list)
+
+    def has_reduction(self):
+        return self.reduce_op is not None
+
+    def fresh_index(self):
+        res = self.count
+        self.count += 1
+        return res
+
+    def fresh_premap_param(self, *args, **kwargs):
+        index = self.fresh_index()
+        var = _FusionVarCUDA(index, *args, **kwargs)
+        self.param_list.append(var)
+        return var
+
+    def fresh_postmap_param(self, *args, **kwargs):
```

Comment:
Prefix with `_` if this method is used only within the class (and not meant to be used from outside).